### PR TITLE
Support site Pulp server for image builds

### PR DESCRIPTION
**Pull request by wtripp180901**
- Added options to allow using a local Pulp server for DNF packages (but still using StackHPC Ark by default).
- Includes an ad-hoc playbook to install/configure Pulp on an existing instance
- Includes an ad-hoc playbook to sync defined Pulp snapshots with Ark.
- Includes configuration to automatically sync Pulp snapshots on image build - this is untested.
- Guards against injecting Pulp credentials (when using Ark directly) into cluster instances.

### CODE REVIEW

**Comment by wtripp180901** [comment]
leafcloud site test: https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/12199387002

**Comment by wtripp180901** [comment]
https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/12201370568

**Comment by wtripp180901** [comment]
https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/12201486557

**Comment by wtripp180901** [comment]
test https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/12281871288

**Comment by wtripp180901** [comment]
https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/12297171159

**Comment by wtripp180901** [comment]
https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/12297434262

**Comment by wtripp180901** [comment]
nightly: https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/12298254618

**Comment by wtripp180901** [comment]
nightly: https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/12298497417

**Comment by wtripp180901** [comment]
nightly https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/12298719350

**Comment by wtripp180901** [comment]
nightly: https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/12298906032

**Comment by wtripp180901** [comment]
nightly: https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/12299607485/job/34326629840
fatimage: https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/12300002950

**Comment by wtripp180901** [comment]
^ this commit was the last tested fatimage before we decided to bypass site pulp for now and point the repos back at ark

File: ansible/fatimage.yml [hunk 1]
```diff
@@ -17,6 +17,25 @@
   import_playbook: "{{ hook_path if hook_path | exists else 'noop.yml' }}"
   when: hook_path | exists
 
+- name: Sync pulp repos with upstream
+  hosts: pulp
```
Comment (sjpb) [inline ansible/fatimage.yml:21]: Note for self: adding hosts to this group JUST controls whether sync happens

File: ansible/roles/pulp_site/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
+pulp_site_url: "{{ appliances_pulp_url }}"
+pulp_site_port: 8080
+pulp_site_username: admin # shouldn't be changed
+pulp_site_upstream_content_url: https://ark.stackhpc.com/pulp/content
+pulp_site_upstream_username: slurm-app-ci
+pulp_site_upstream_password: "{{ lookup('ansible.builtin.env', 'ARK_PASSWORD') }}"
```
Comment (sjpb) [inline ansible/roles/pulp_site/defaults/main.yml:6]: Nope, I don't think we should look this up from env. It should normally be set as vaulted group_vars, so should be empty here.

Maybe we could help signpost this by adding:

```yaml
# environments/common/inventory/group_vars/pulp_site:
# Add your creds for stackhpc ark here
# pulp_site_upstream_username:
# pulp_site_upstream_password: 
```

File: ansible/roles/pulp_site/defaults/main.yml [hunk 3]
```diff
@@ -0,0 +1,43 @@
+pulp_site_url: "{{ appliances_pulp_url }}"
+pulp_site_port: 8080
+pulp_site_username: admin # shouldn't be changed
+pulp_site_upstream_content_url: https://ark.stackhpc.com/pulp/content
+pulp_site_upstream_username: slurm-app-ci
```
Comment (sjpb) [inline ansible/roles/pulp_site/defaults/main.yml:5]: Nope, there is no default for this really.

File: ansible/roles/pulp_site/defaults/main.yml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+pulp_site_upstream_password: "{{ lookup('ansible.builtin.env', 'ARK_PASSWORD') }}"
+_pulp_site_rocky_prefix: "{{ pulp_site_target_distribution }}/{{ pulp_site_target_distribution_version }}"
+pulp_site_default_upstream_suffix: "{{ pulp_site_target_arch }}/os"
+pulp_site_password: "{{ vault_pulp_admin_password }}"
```
Comment (sjpb) [inline ansible/roles/pulp_site/defaults/main.yml:9]: Can we move this to below `pulp_site_username` to make it clear its the username for that password?

File: ansible/roles/pulp_site/defaults/main.yml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+pulp_site_validate_certs: false
+pulp_site_install_dir: '/home/rocky/pulp'
+pulp_site_selinux_suffix: "{{ ':Z' if ansible_selinux.status == 'enabled' else '' }}"
+pulp_site_target_facts: "{{ hostvars[groups['builder'][0]]['ansible_facts'] }}"
+pulp_site_target_arch: "{{ pulp_site_target_facts['architecture'] }}"
+pulp_site_target_distribution: "{{ pulp_site_target_facts['distribution'] | lower }}"
+pulp_site_target_distribution_version: "{{ pulp_site_target_facts['distribution_version'] }}"
+pulp_site_target_distribution_version_major: "{{ pulp_site_target_facts['distribution_major_version'] }}"
+pulp_site_version_timestamps: "{{ appliances_repo_minor_timestamps[pulp_site_target_distribution_version] }}"
+pulp_site_major_version_timestamps: "{{ appliances_repo_major_timestamps[pulp_site_target_distribution_version_major] }}"
```
Comment (sjpb) [inline ansible/roles/pulp_site/defaults/main.yml:19]: Discussed elsewhere to ditch major/minor bit

File: ansible/adhoc/deploy-pulp.yml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  tasks:
+  - ansible.builtin.add_host:
+      name: "{{ pulp_server }}"
+      group: "_pulp_host"
+
+- name: Install pulp on server and add to config
+  become: yes
+  hosts: _pulp_host
+  tasks:
+
```
Comment (sjpb) [inline ansible/adhoc/deploy-pulp.yml:14]:
```suggestion
```

File: ansible/adhoc/deploy-pulp.yml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+- name: Add temporary pulp server host
+  hosts: localhost
+  tasks:
+  - ansible.builtin.add_host:
+      name: "{{ pulp_server }}"
+      group: "_pulp_host"
+
+- name: Install pulp on server and add to config
+  become: yes
+  hosts: _pulp_host
```
Comment (sjpb) [inline ansible/adhoc/deploy-pulp.yml:12]: I'd be keen to add a specific pulp user here, lets leave that for the moment though

File: environments/.stackhpc/inventory/group_vars/builder.yml [hunk 8]
```diff
@@ -1 +1,14 @@
...
+# Uncomment below to use CI pulp servers
+
+# pulp_server_config:
+#   LEAFCLOUD:
+#     url: http://192.168.10.157:8080
+#     password: lookup('env','LEAFCLOUD_PULP_PASSWORD')
+
+# appliances_pulp_url: "{{ pulp_server_config[lookup('env','CI_CLOUD')].url }}"
+# pulp_site_password: "{{ pulp_server_config[lookup('env','CI_CLOUD')].password }}"
+
```
Comment (sjpb) [inline environments/.stackhpc/inventory/group_vars/builder.yml:12]:
```suggestion
# Configure to use ark directly:
```

File: environments/common/inventory/groups [hunk 9]
```diff
@@ -144,3 +144,10 @@ freeipa_client
 
 [lustre]
 # Hosts to run lustre client
+
+[dnf_repos:children]
+# Hosts to replace system repos with Pulp repos
+builder
+
+[pulp:children]
+# Hosts used to run Pulp API commands
```
Comment (sjpb) [inline environments/common/inventory/groups:153]:
```suggestion
# Add builder to this group to enable automatically syncing of pulp during image build
```

Comment (sjpb) [reply environments/common/inventory/groups:153]: Will can you update the docs if necessary to be more in line with this comment pls?

**Review by sjpb — CHANGES_REQUESTED** [review]

File: ansible/roles/pulp_site/defaults/main.yml (hunk 2 above)
Comment (wtripp180901) [reply ansible/roles/pulp_site/defaults/main.yml:6]: this is to get the gh secret from CI but yeah will move to .stackhpc

File: ansible/roles/pulp_site/defaults/main.yml (hunk 3 above)
Comment (wtripp180901) [reply ansible/roles/pulp_site/defaults/main.yml:5]: CI username, will move to .stackhpc

File: environments/.stackhpc/inventory/group_vars/builder.yml (hunk 8 above)
Comment (wtripp180901) [reply environments/.stackhpc/inventory/group_vars/builder.yml:12]:
```suggestion
# Alternatively, configure to use ark directly:
```

**Comment by wtripp180901** [comment]
https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/12351955189

**Comment by wtripp180901** [comment]
https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/12356348160

**Comment by wtripp180901** [comment]
https://github.com/stackhpc/ansible-slurm-appliance/actions/runs/12356576826

**Review by sjpb — CHANGES_REQUESTED** [review]

**Review by sjpb — APPROVED** [review]
LGTM

**Pull request merged**